client/blb/client.go

```diff
@@ -1127,109 +1126,123 @@ func (cli *Client) readOneTractReplicated(
+		// phase.
+		readFuncBackup := func(wg *sync.WaitGroup, n int, thisBLoc []byte) {
+			defer wg.Done()
+			host := tract.Hosts[order[n]]
 			if host == "" {
 				log.V(1).Infof("read %s from tsid %d: no host", tract.Tract, tract.TSIDs[n])
-				ch <- tractResult{0, 0, err, badVersionHost}
+				ch <- tractResultRepl{
+					tractResult: tractResult{0, 0, err, badVersionHost},
+					thisB:       thisBLoc,
+				}
 				return
 			}
 
-			// TODO(eric): fill in otherHosts when ts-ts cancellation is done.
-			read, err := cli.tractservers.ReadInto(ctx, host, nil, reqID, tract.Tract, tract.Version, thisB, thisOffset)
-			if err == core.ErrVersionMismatch {
-				badVersionHost = host
-			}
-			if err != core.NoError && err != core.ErrEOF {
-				// TODO(eric): redo bad TS reporting mechanism.
-				ch <- tractResult{0, 0, err, badVersionHost}
-				return
-			}
-			log.V(1).Infof("read %s from tractserver at address %s: %s", tract.Tract, host, err)
-			// In case of short read, i.e., read < len(thisB), we should
-			// pad the result with 0's. This doesn't need to be done
-			// explicitly if 'thisB' given to us starts to be empty.
-			// However, just in case it's not, we will ensure that by
-			// overwritting untouched bytes with 0's.
-			for i := read; i < len(thisB); i++ {
-				thisB[i] = 0
+			result := readFunc(host, n, thisBLoc)
+
+			ch <- tractResultRepl{
+				tractResult: result,
+				badReadHost: host,
+				thisB:       thisBLoc,
 			}
-			ch <- tractResult{len(thisB), read, err, badVersionHost}
-			return
 		}
 
-		// Backup request logic involves the following:
-		// 1) Send the first read to the first host (item 0 available from the orderCh) in a goroutine.
-		// 2) Start another goroutine to issue backup requests. Each subsequent request is spawned after the
-		// configured backup time.After delay time expires.
-		// 3) Block until the first read is returned from the goroutines servicing the reads. Note that if the
-		// the request returned fails, we accept that and continume on to the fallback phase for sequential reads.
-		// We could optimize this later and scan for the first successful backup request, but that may not be
-		// worth the latency savings.
+		var wg sync.WaitGroup
+		nb := min(maxNumBackups+1, len(order))
+		wg.Add(nb)
 		go func() {
-			nb := min(maxNumBackups+1, len(order))
-			for i := 0; i < nb; i++ {
+			for n := 0; n < nb; n++ {
 				select {
-				case <-cli.backupReadState.backupDelayFunc(time.Duration(i) * delay):
-					go readFunc(order[i])
+				case <-cli.backupReadState.delayFunc(time.Duration(n) * delay):
+					thisBLoc := make([]byte, len(thisB))
+					go readFuncBackup(&wg, n, thisBLoc)
 				case <-ctx.Done():
-					log.V(1).Infof("backup read %s-%d to tractserver cancelled", tract.Tract, i)
+					log.V(1).Infof("backup read %s-%d to tractserver cancelled", tract.Tract, n)
 					break
 				}
 			}
 		}()
+		go func() {
```

Comment:
This is kind of an unfortunate pattern, though sometimes necessary.

Just thinking out loud.. what if you ditched the waitgroup and just had each reader goroutine promise to write something to ch, then read exactly that number of things?

Reply:
Some of the go docs/book/blogs got me thinking this way :/ sure I can ditch the waitgroup. I think sending a fixed number of things and reading a fixed number of things sounds cleaner.  I take it that doing a cancel() on the first good read seen will stop the others as mentioned below, but in that case I am thinking that we should still have the cancelled goroutines send some kind of error in the `tractResultRepl`. That way the readers fulfill their promise to always send something and we won't deadlock in the loop that reads from the channel.